File: 2663. Lexicographically Smallest Beautiful String/Solution.cpp
class Solution {
public:
    string smallestBeautifulString(string s, int k) {
        int n = s.length(), i = n - 1;
        while (i >= 0) {
            s[i]++;
            if (s[i] == char('a' + k)) {
                i--;
            } else if ((i - 1 < 0 || s[i - 1] != s[i]) && (i - 2 < 0 || s[i - 2] != s[i])) {
                for (int j = i + 1; j < n; j++) {
                    vector<bool> can(3, true);
                    for (int l = max(0, j - 2); l < j; l++) {
                        if (s[l] - 'a' > 2) continue;
                        can[s[l] - 'a'] = false;
                    }
                    for (int l = 0; l < 3; l++) {
                        if (can[l] == true) {
                            s[j] = char(l + 'a');
                            break;
                        }
                    }
                }
                return s;
            }
        }
        return "";
    }
};

File: 2663. Lexicographically Smallest Beautiful String/Solution.py
class Solution:
    def smallestBeautifulString(self, s: str, k: int) -> str:
        ch = [ord(c) - ord('a') for c in s]
        n, i = len(ch), len(ch) - 1
        while i >= 0:
            ch[i] += 1
            if ch[i] == k:
                i -= 1
            elif (i - 1 < 0 or ch[i - 1] != ch[i]) and (i - 2 < 0 or ch[i - 2] != ch[i]):
                for j in range(i + 1, n):
                    can = [True, True, True]
                    for k in range(max(0, j - 2), j):
                        if ch[k] > 2: continue
                        can[ch[k]] = False
                    for k in range(3):
                        if can[k]:
                            ch[j] = k
                            break
                return ''.join(chr(ord('a') + c) for c in ch)
        return ''